rosbag2_transport/test/rosbag2_transport/test_record_all_include_unpublished_topics.cpp

```diff
@@ -0,0 +1,201 @@
+// Copyright 2022 Seegrid Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "test_msgs/msg/basic_types.hpp"
+#include "test_msgs/message_fixtures.hpp"
+
+#include "rosbag2_test_common/publication_manager.hpp"
+#include "rosbag2_test_common/wait_for.hpp"
+
+#include "rosbag2_transport/recorder.hpp"
+
+#include "record_integration_fixture.hpp"
+
+using namespace std::chrono_literals;  // NOLINT
+
+TEST_F(RecordIntegrationTestFixture, record_all_include_unpublished_false_ignores_unpublished)
+{
+  std::string string_topic = "/string_topic";
+  auto node = std::make_shared<rclcpp::Node>("test_string_msg_listener_node");
+  auto string_msgs_sub = node->create_subscription<test_msgs::msg::Strings>(
+    string_topic, 10, [](test_msgs::msg::Strings::ConstSharedPtr) {});
+
+  rosbag2_transport::RecordOptions record_options = {true, false, {}, "rmw_format", 100ms};
+  record_options.include_unpublished_topics = false;
+  auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    std::move(writer_), storage_options_, record_options);
+  recorder->record();
+
+  start_async_spin(recorder);
+
+  auto & writer = recorder->get_writer_handle();
+  MockSequentialWriter & mock_writer =
+    static_cast<MockSequentialWriter &>(writer.get_implementation_handle());
+
+  size_t expected_messages = 0;
+  rosbag2_test_common::wait_until_shutdown(
+    std::chrono::seconds(2),
+    [&mock_writer, &expected_messages]() {
+      return mock_writer.get_messages().size() > expected_messages;
+    });
+  (void) expected_messages;  // we can't say anything here, there might be some rosout
```

Comment:
@theseankelly hmm, if the expectation below is to succeed, no messages should have been received

Reply:
Also, why not constrain recording to relevant topics to mitigate cross-talk?

Reply:
That particular line (59) was copied from other tests in the same suite  (see `test_record_all_no_discovery.cpp` for example). I'm not crazy about this pattern but I thought consistency would be better in this case.  As an alternative I could just remove the `expected_messages` variable altogether. In fact, I should do so.

The reason not to constrain to relevant topics is because this test suite is a test of the `-a` flag

Reply:
Actually I'm glad you brought this up because I misunderstood the usage of `rosbag2_test_common::wait_until_shutdown` -- I think some of my other tests may be ending prematurely. Stay tuned...